File: c++/topological_sort.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

vector<int>adj[501];
void topological_sort() {
	int v, e;
	cin >> v >> e;
	vector<int>indegree(v + 1, 0);
	for (int i = 0; i < e; i++) {
		int a, b;
		cin >> a >> b;
		adj[a].push_back(b);
		indegree[b]++;
	}
	queue<int>q;
	vector<int>route;
	for (int i = 1; i <= v; i++)
		if (indegree[i] == 0)
			q.push(i);

	while (!q.empty()) {
		int here = q.front();
		q.pop();
		route.push_back(here);
		for (int i = 0; i < adj[here].size(); i++) {
			int there = adj[here][i];
			indegree[there]--;
			if (indegree[there] == 0)
				q.push(there);
		}
	}
	for (int i = 0; i < route.size(); i++)
		cout << route[i] << " ";
}

int main() {
	topological_sort();
}

File: python/topological_sort.py
from collections import deque

v,e=map(int,input().split())
indegree= [0]*(v+1)
adj=[[] for _ in range(v+1)]
for _ in range(e):
    a,b=map(int,input().split())
    indegree[b]+=1
    adj[a].append(b)

def topological_sort():
    result=[]
    q=deque()
    for i in range(1,v+1):
        if indegree[i]==0:
            q.append(i)
    
    while q:
        a = q.popleft()
        result.append(a)
        for i in adj[a]:
            indegree[i]-=1
            if indegree[i]==0:
                q.append(i)
    for i in result:
        print(i,end=' ')
topological_sort()
